Add unified tagging advance usage guide (#7597)
DataDog/documentation: content/en/developers/guide/unified-tagging-advanced-usage.md

---
title: Unified Tagging Advanced Usage Guide
kind: guide
aliases:
  - /developers/guide/unified-tagging-advanced-usage
---

## Overview

> This guide suggests ways to configure and migrate to [unified service tagging][1] based on specific use cases.

Author: Note: in the process of updating intro 🦖

Reviewer: What were you thinking of adding? I think this works on its own?

Author: It seemed a bit short, but going back to the content I was like eh, this actually works best haha. Ignore my 🦖 we are 👍

**Note**: While some products support arbitrary tags, others have more specific expectations. As such, navigation between products might be difficult if one data source has a tag that another does not have or does not support.

### Containerized Environments

#### Metrics

Since tags can be appended to data points, you have a lot of freedom in setting desired tags. Autodiscovered tags are automatically added to all metrics collected.

#### APM

`env` and `service` are core tags in APM, so it is not possible to replace them with differently named tags. However, APM does allow for data to be [aggregated along more primary tags] than just `env`. Since host tags are added to traces and trace metrics, they can be used as well (e.g., availability-zone).

Trace metrics also have a specific [whitelist of tags][2]. Container tags are not included.

Autodiscovered tags associated with containers are added to `container_info` in span metadata.

#### Logs

> Similar to APM, service is a core tag that is used to help organize logs data. Additionally, it is not possible to link from a log to the related APM service without it.

Author: ```suggestion

Similar to APM, service is a core tag that is used to help organize logs data. Additionally, it is not possible to link from a log to the related APM service without it.
```

Similar to metrics, however, Autodiscovered tags for the container as well as host tags for the agent are added to all logs.

You also have the ability to add custom fields to your logs in-code, which can be mapped to tags or attributes downstream in a [Datadog log processing pipeline][3].

## Using Standard Labels

We recommend using both standard labels and environment variables. However, standard labels can be seen as an alternative to using environment variables, particularly for applications that will not benefit from using those variables in their runtime. 3rd party software like Redis, MySQL, and Nginx are a few examples. Since these services still generate infrastructure metrics and data from integration checks, it's valuable to have all of that data tagged with `env`, `service`, and `version`.

> If you would like to tag Kubernetes state metrics with `env`, `service`, and `version` then the standard labels offer the easiest way. The `DD environment variables of a container cannot be used by the Agent to tag these metrics, so the labels are a more natural option.

Author: ```suggestion
If you would like to tag Kubernetes state metrics with `env`, `service`, and `version` then the standard labels offer the easiest way. The `DD` environment variables of a container cannot be used by the Agent to tag these metrics, so the labels are a more natural option.
```

### Declaring environment as a label

It's important to declare `env `as a label even if it's already defined in your agent. Configuring `env` closer to the source of the data (e.g., APM traces, logs) helps to avoid inconsistencies where the Agent `env` might be different. Making `env` part of the service's configuration guarantees a service-centric source of truth.

### Using standard labels with existing Kubernetes tags annotation

> Kubernetes users can continue to use these general tags. However, using the specific labels has a few benefits:

Author: ```suggestion

Kubernetes users can continue to use these general tags. However, using the specific labels has a few benefits:
```

- You can directly reference them for environment variable injection via the Kubernetes downward API.
- The service standard label can simplify the definition of service for logs.

> ## Using standard labels for specific containers

Author: ```suggestion
### Using standard labels for specific containers
```

> Since the `DD` environment variables are injected at the container level, they can differ from container to container. However, if you want to use the standard labels as well for specific containers, then you will need to use their container-specific variants:

Author: ```suggestion

Since the `DD` environment variables are injected at the container level, they can differ from container to container. However, if you want to use the standard labels as well for specific containers, then you will need to use their container-specific variants:
```

---
title: Unified Tagging Advanced Usage Guide
kind: guide
aliases:
  - /developers/guide/unified-tagging-advanced-usage
---

## Overview

This guide suggests ways to configure and migrate to [unified service tagging][1] based on specific use cases.

## Using Custom Tags

You can continue to use the `env` `service` and `version` tags as they are configured for unified service tagging. However, if you'd like to achieve a unified tagging experience with your own custom tags, there are some options available to you:

> **Note**: While some products support arbitrary tags, others have more specific expectations. As such, navigation between products might be difficult if one data source has a tag that another does not have or does not support.

Reviewer: Drop the `Note` per Yair's feedback on the other PR?

Author: His note was pertaining to having a note at the start of a section right below the header, which makes sense. It's a little odd to start a section with a note. However, I think this note make sense here as it is something we want people to make note of before going through these options.

> **Note**: While some products support arbitrary tags, others have more specific expectations. As such, navigation between products might be difficult if one data source has a tag that another does not have or does not support.

Reviewer 2: It's a little confusing that there is a trailing `:` at the end of this sentence. Maybe rephrase as `there are several options available which are listed below.` or similar. @sarina-dd @jdgumz thoughts?

> ### Containerized Environments

Author: ```suggestion
### Containerized environments
```

#### Metrics

Since tags can be appended to data points, you have a lot of freedom in setting desired tags. Autodiscovered tags are automatically added to all metrics collected.

#### APM

> `env` and `service` are core tags in APM, so it is not possible to replace them with differently named tags. However, APM does allow for data to be [aggregated along more primary tags] than just `env`. Since host tags are added to traces and trace metrics, they can be used as well (e.g., availability-zone).

Reviewer: Missing a link to the setting primary tags to scope page? Looks like we'd want one for `[aggregated along more primary tags]`.

> `env` and `service` are core tags in APM, so it is not possible to replace them with differently named tags. However, APM does allow for data to be [aggregated along more primary tags] than just `env`. Since host tags are added to traces and trace metrics, they can be used as well (e.g., availability-zone).

Author: ```suggestion
`env` and `service` are core tags in APM, so it is not possible to replace them with differently named tags. However, APM does allow for data to be [aggregated along more primary tags][4] than just `env`. Since host tags are added to traces and trace metrics, they can be used as well (e.g., availability-zone).
```

> Trace metrics also have a specific [whitelist of tags][2]. Container tags are not included.

Reviewer: Maybe we can move the bit about container tags to the next sentence? Or combine these two sentences? 

Something like:
> Autodiscovered tags associated with containers are added to `container_info` in span metadata.
> However, these container tags are not part of the whitelisted tags for trace metrics.

> Trace metrics also have a specific [whitelist of tags][2]. Container tags are not included.

Author: ```suggestion
Autodiscovered tags associated with containers are added to `container_info` in span metadata. However, these container tags are not part of the [whitelisted tags][2] for trace metrics.
```

> Autodiscovered tags associated with containers are added to `container_info` in span metadata.

Author: ```suggestion
```

#### Logs

> Similar to APM, service is a core tag that is used to help organize logs data. Additionally, it is not possible to link from a log to the related APM service without it.

Reviewer 2: Highlight `service` in backticks here?

Similar to metrics, however, Autodiscovered tags for the container as well as host tags for the agent are added to all logs.

You also have the ability to add custom fields to your logs in-code, which can be mapped to tags or attributes downstream in a [Datadog log processing pipeline][3].

## Using Standard Labels

We recommend using both standard labels and environment variables. However, standard labels can be seen as an alternative to using environment variables, particularly for applications that will not benefit from using those variables in their runtime. 3rd party software like Redis, MySQL, and Nginx are a few examples. Since these services still generate infrastructure metrics and data from integration checks, it's valuable to have all of that data tagged with `env`, `service`, and `version`.

If you would like to tag Kubernetes state metrics with `env`, `service`, and `version` then the standard labels offer the easiest way. The `DD` environment variables of a container cannot be used by the Agent to tag these metrics, so the labels are a more natural option.

### Declaring environment as a label

> It's important to declare `env `as a label even if it's already defined in your agent. Configuring `env` closer to the source of the data (e.g., APM traces, logs) helps to avoid inconsistencies where the Agent `env` might be different. Making `env` part of the service's configuration guarantees a service-centric source of truth.

Reviewer: I'm wondering if we can remove the first sentence and just start with
> Configuring `env` closer to the ...

Starting with the description of an edge case might be confusing. We could rephrase it into an example for why this usage pattern is helpful.

> It's important to declare `env `as a label even if it's already defined in your agent. Configuring `env` closer to the source of the data (e.g., APM traces, logs) helps to avoid inconsistencies where the Agent `env` might be different. Making `env` part of the service's configuration guarantees a service-centric source of truth.

Author: ```suggestion
Configuring `env` closer to the source of the data (e.g., APM traces, logs) helps to avoid inconsistencies where the Agent `env` might be different. Making `env` part of the service's configuration guarantees a service-centric source of truth.
```

### Using standard labels with existing Kubernetes tags annotation

Kubernetes users can continue to use these general tags. However, using the specific labels has a few benefits:

- You can directly reference them for environment variable injection via the Kubernetes downward API.
- The service standard label can simplify the definition of service for logs.

### Using standard labels for specific containers

Since the `DD` environment variables are injected at the container level, they can differ from container to container. However, if you want to use the standard labels as well for specific containers, then you will need to use their container-specific variants:

```yaml
tags.datadoghq.com/<container>.env
tags.datadoghq.com/<container>.service
tags.datadoghq.com/<container>.version
```

[1]: /getting_started/tagging/unified_service_tagging
[2]: /metrics/distributions/#customize-tagging

> [3]: /logs/processing/pipelines/

Author: ```suggestion
[3]: /logs/processing/pipelines/
[4]: /tracing/guide/setting_primary_tags_to_scope/
```

---
title: Unified Tagging Advanced Usage Guide
kind: guide
aliases:
  - /developers/faq/unified-tagging-advanced-usage
---

## Overview

This guide suggests ways to configure and migrate to [unified service tagging][1] based on specific use cases.

## Using Standard Labels

The standard labels can be seen as an alternative to using environment variables, particularly for applications that will not benefit from using those variables in their runtime. 3rd party software like Redis, MySQL, and Nginx are a few examples. Since these services still generate infrastructure metrics and data from integration checks, it's valuable to have all of that data tagged with `env`, `service`, and `version`.

If you do not use Docker as your container runtime on Kubernetes then you will have to use the standard labels to tag Kubernetes metrics. The Datadog Agent does not yet support the extraction of environment variables as tags from container runtimes outside of Docker.

> Lastly, if you would like to tag Kubernetes state metrics with `env`, `service`, and `version` then the standard labels offer the easiest way. The DD environment variables of a container cannot be used by the Agent to tag these metrics, so the labels are a more natural option.

Reviewer: The agent can map env vars to tags if Docker is the container runtime. This is why we can recommend using just env vars for Docker generally.

We endorse standard labels in Kubernetes since users might have container runtimes other than Docker (e.g., `containerd`).

```yaml
tags.datadoghq.com/<container>.env
tags.datadoghq.com/<container>.service
tags.datadoghq.com/<container>.version
```

## What if I want/have different tags for env/service/version?
You can continue to use these tags as you have them configured today. We have standardized on env, service, and version to make tagging consistent and simple. If you'd like to achieve a unified tagging experience with your own custom tags, there are some options available to you.

But note that while some products support arbitrary tags, others have more specific expectations. As such, navigation between products might be difficult if one data source has a tag that another does not have or does not support.

Why do I see a different version in my spans if I've set DD_VERSION?
You might be hard-coding the version tag in your spans. To address this issue, rename the span tag to something else (e.g., payload_version). That way it won't conflict with setting the service's version for the version tag.

> ### Containerized environments

Reviewer 3: overall, I think it's better to include this here. it seems pretty useful and it never hurts to state the same thing in different ways and in different places

#### APM

`env` and `service` are core tags in APM, so it is not possible to replace them with differently named tags. However, APM does allow for data to be [aggregated along more primary tags] than just `env`. Since host tags are added to traces and trace metrics, they can be used as well (e.g., availability-zone).

Trace metrics also have a specific [whitelist of tags]. Container tags are not included.

Autodiscovered tags associated with containers are added to `container_info` in span metadata.

#### Logs
Similar to APM, service is a core tag that is used to help organize logs data. Additionally, it is not possible to link from a log to the related APM service without it.

Similar to metrics however, Autodiscovered tags for the container as well as host tags for the agent are added to all logs.

You also have the ability to add custom fields to your logs in-code, which can be mapped to tags or attributes downstream in a [Datadog log processing pipeline].

### Non-containerized

While host tags can be used to conveniently add service to all data sources, we recommend against setting it as a host tag - especially if you have multiple APM services that run on a host.

> We'd recommend taking advantage of each product's options for tagging.

Reviewer: @apigirl @cswatt 

If we keep these sections to talk about options for tagging, then we'll want to include links to the appropriate docs by product.

Although the direction/guidance is largely the same across containerized and non-containerized. The only big difference is that we don't have autodiscovery in the latter.

Should we re-org to discuss containerized vs. non-containerized and then have individual sections by product?
